# The idea of this file is that all the commonly-used scripts are in random,
# disparate locations. In order to have all the scripts in one place, we will 
# symlink the scripts to $(SCRIPTS_DIR) and run them from there. 
# 
# When it's time to package up the analysis, we will simply run tar using the -h
# option to follow symlinks. 
# 
# Remember to declare the script as a dependency of the file! 

# This directory is a subdirectory of the main directory
SCRIPTS_DIR = scripts

# File locations

ifeq ($(UNAME),Linux)
HOME :=  /lab/bartel1_ata/koppstein
endif
ifeq ($(UNAME),Darwin)
HOME := /Users/dnk63
endif
TAK_ONLY = $(HOME)/Tak_Only
DKSOFTWARE_DIR = $(HOME)/science/software/dklib
INFLUENZA_SCRIPTS_DIR = $(DKSOFTWARE_DIR)/influenza
DKLIB_SCRIPTS = $(DKSOFTWARE_DIR)/scripts

########### LINK SCRIPTS #############

ALL_SCRIPTS := UNZIP_FASTQ COUNT_TRANSCRIPTS DISCARD_FASTQ FASTQ_QUALITY_FILTER COLLAPSE_PAR TSS_SCRIPT PLOT_PAR PLOT_LENGTH_DISTRIBUTION SHUFFLE_SCRIPT PROCESS_PET OFFSET_BED CUSTOM_COLLAPSER EQUALIZE_LENGTHS FIND_CLOSEST_MAPPING UPPERCASE_FASTQ MAPPING_TO_BED HISTOGRAM_COLUMN PLOT_PAR_AND_LENGTH LENGTH_VS_PAR FASTX_TO_GTF BED_WITH_NO_GS SPLIT_AND_MAP CUSTOM_BEDTOOLS_INTERSECT GRAPH_OF_BEDTOOLS_INTERSECT DEPARSE_BEDTOOLS_OUTPUT INTERSECT_SEQRECORDS INTERSECT_TABLES CUSTOM_BAM_TO_HUMAN_COORDS BAM_TO_TABLE PREPROCESS_BEDTOOLS_OUTPUT THREE_PRIME_OF_CLEAVAGE_PLOT PREPROCESS_TABLE_FOR_THREEPRIME_OF_CLEAVAGE PLOT_NUCLEOTIDE_DIST PSSM_PLOT SUBSET_TABLE DEPYTHONTEX PYTHONTEX COMPARE_DATA MAKE_SUPPLEMENTAL_TABLE PRIME_AND_REALIGN QUICK_COUNT_NON_CAAAAGCAG_TRIMMED COLLAPSE_PAR_V2 FASTQ_TO_TABLE PLOT_LENGTH_DISTRIBUTIONS PSSM_PLOTS U1_VS_U2 LINEAR_REGRESSION PLOT_PAR_LENGTHS FIND_DINUCLEOTIDE_AT_JUNCTION CONVERT_FORMATS PLOT_DISTS PLOT_EACH_LENGTH_DISTRIBUTION PLOT_U_SNRNAS NOG_SCRIPT NOG_STATS_COMBINER CONVERT_GTF_NAMES TOP_GENES FIND_DINUCLEOTIDE_AT_JUNCTION_WITH_BACKGROUND TRIM_SEQUENCES

UNZIP_FASTQ = $(SCRIPTS_DIR)/unzip_fastq.py
UNZIP_FASTQ_LOCATION = $(DKLIB_SCRIPTS)/unzip_fastq.py

COUNT_TRANSCRIPTS_LOCATION = \
$(HOME)/science/software/count_transcripts/count_transcripts.py
COUNT_TRANSCRIPTS = $(SCRIPTS_DIR)/count_transcripts.py

EXTRACT_FASTA_LOCATION = $(DKLIB_SCRIPTS)/extract_fasta.py
EXTRACT_FASTA = $(SCRIPTS_DIR)/extract_fasta.py

DISCARD_FASTQ_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/discard_fastq.py
DISCARD_FASTQ = $(SCRIPTS_DIR)/discard_fastq.py

FASTQ_QUALITY_FILTER_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/fastq_quality_filter.py
FASTQ_QUALITY_FILTER = $(SCRIPTS_DIR)/fastq_quality_filter.py

COLLAPSE_PAR = $(SCRIPTS_DIR)/collapse_par.py
COLLAPSE_PAR_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/collapse_par.py

COLLAPSE_PAR_V2 = $(SCRIPTS_DIR)/collapse_par_v2.py
COLLAPSE_PAR_V2_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/collapse_par_v2.py


TSS_SCRIPT_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/find_all_starts.py
TSS_SCRIPT = $(SCRIPTS_DIR)/find_all_starts.py

PLOT_PAR_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/plot_par.R
PLOT_PAR = $(SCRIPTS_DIR)/plot_par.R

PLOT_LENGTH_DISTRIBUTION_LOCATION = $(DKLIB_SCRIPTS)/plot_length_distribution.py
PLOT_LENGTH_DISTRIBUTION = $(INFLUENZA_SCRIPTS_DIR)/plot_length_distribution.py

SHUFFLE_SCRIPT_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/shuffle.py
SHUFFLE_SCRIPT = $(SCRIPTS_DIR)/shuffle.py

PROCESS_PET_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/process_pet.py
PROCESS_PET = $(SCRIPTS_DIR)/process_pet.py

OFFSET_BED_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/offset_bed.py
OFFSET_BED = $(SCRIPTS_DIR)/offset_bed.py

CUSTOM_COLLAPSER_LOCATION = $(DKLIB_SCRIPTS)/custom_collapser.py
CUSTOM_COLLAPSER = $(SCRIPTS_DIR)/custom_collapser.py

EQUALIZE_LENGTHS_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/equalize_lengths.py
EQUALIZE_LENGTHS = $(SCRIPTS_DIR)/equalize_lengths.py

FIND_CLOSEST_MAPPING = $(SCRIPTS_DIR)/find_closest_mapping.py
FIND_CLOSEST_MAPPING_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/find_closest_mapping.py

UPPERCASE_FASTQ = $(SCRIPTS_DIR)/uppercase_fastq.py
UPPERCASE_FASTQ_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/uppercase_fastq.py

MAPPING_TO_BED = $(SCRIPTS_DIR)/mapping_to_bed.py
MAPPING_TO_BED_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/mapping_to_bed.py

HISTOGRAM_COLUMN = $(SCRIPTS_DIR)/histogram_column.R
HISTOGRAM_COLUMN_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/histogram_column.R

PLOT_PAR_AND_LENGTH = $(SCRIPTS_DIR)/plot_par_and_length.R
PLOT_PAR_AND_LENGTH_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/plot_par_and_length.R

LENGTH_VS_PAR = $(SCRIPTS_DIR)/length_vs_par.py
LENGTH_VS_PAR_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/length_vs_par.py

FASTX_TO_GTF = $(SCRIPTS_DIR)/fastx_to_gtf.py
FASTX_TO_GTF_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/fastx_to_gtf.py

BED_WITH_NO_GS = $(SCRIPTS_DIR)/bed_with_no_gs.py
BED_WITH_NO_GS_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/bed_with_no_gs.py

SPLIT_AND_MAP = $(SCRIPTS_DIR)/split_and_map.py
SPLIT_AND_MAP_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/split_and_map.py

CUSTOM_BEDTOOLS_INTERSECT = $(SCRIPTS_DIR)/custom_bedtools_intersect.py
CUSTOM_BEDTOOLS_INTERSECT_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/custom_bedtools_intersect.py

GRAPH_OF_BEDTOOLS_INTERSECT = $(SCRIPTS_DIR)/graph_of_bedtools_intersect.py
GRAPH_OF_BEDTOOLS_INTERSECT_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/graph_of_bedtools_intersect.py

DEPARSE_BEDTOOLS_OUTPUT = $(SCRIPTS_DIR)/deparse_bedtools_output.py
DEPARSE_BEDTOOLS_OUTPUT_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/deparse_bedtools_output.py

INTERSECT_SEQRECORDS = $(SCRIPTS_DIR)/intersect_seqrecords.py
INTERSECT_SEQRECORDS_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/intersect_seqrecords.py

CUSTOM_BAM_TO_HUMAN_COORDS = $(SCRIPTS_DIR)/custom_bam_to_human_coords.py
CUSTOM_BAM_TO_HUMAN_COORDS_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/custom_bam_to_human_coords.py

INTERSECT_TABLES = $(SCRIPTS_DIR)/intersect_tables.py
INTERSECT_TABLES_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/intersect_tables.py

BAM_TO_TABLE = $(SCRIPTS_DIR)/bam_to_table.py
BAM_TO_TABLE_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/bam_to_table.py

PREPROCESS_BEDTOOLS_OUTPUT = $(SCRIPTS_DIR)/preprocess_bedtools_output.py
PREPROCESS_BEDTOOLS_OUTPUT_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/preprocess_bedtools_output.py

THREE_PRIME_OF_CLEAVAGE_PLOT = $(SCRIPTS_DIR)/three_prime_of_cleavage_plot.py
THREE_PRIME_OF_CLEAVAGE_PLOT_LOCATION = \
$(INFLUENZA_SCRIPTS_DIR)/three_prime_of_cleavage_plot.py

PREPROCESS_TABLE_FOR_THREEPRIME_OF_CLEAVAGE = $(SCRIPTS_DIR)/preprocess_table_for_threeprime_of_cleavage.py
PREPROCESS_TABLE_FOR_THREEPRIME_OF_CLEAVAGE_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/preprocess_table_for_threeprime_of_cleavage.py

PLOT_NUCLEOTIDE_DIST = $(SCRIPTS_DIR)/plot_nucleotide_dist.py
PLOT_NUCLEOTIDE_DIST_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/plot_nucleotide_dist.py

PSSM_PLOT = $(SCRIPTS_DIR)/pssm_plot.py
PSSM_PLOT_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/pssm_plot.py

SUBSET_TABLE = $(SCRIPTS_DIR)/subset_table.py
SUBSET_TABLE_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/subset_table.py

DEPYTHONTEX = $(SCRIPTS_DIR)/depythontex.py
DEPYTHONTEX_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/depythontex.py

PYTHONTEX = $(SCRIPTS_DIR)/pythontex.py
PYTHONTEX_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/pythontex.py

COMPARE_DATA = $(SCRIPTS_DIR)/compare_data.py
COMPARE_DATA_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/compare_data.py

MAKE_SUPPLEMENTAL_TABLE = $(SCRIPTS_DIR)/make_supplemental_data.py
MAKE_SUPPLEMENTAL_TABLE_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/make_supplemental_table.py

PRIME_AND_REALIGN = $(SCRIPTS_DIR)/prime_and_realign.py
PRIME_AND_REALIGN_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/prime_and_realign.py

QUICK_COUNT_NON_CAAAAGCAG_TRIMMED = $(SCRIPTS_DIR)/quick_count_non_CAAAAGCAG_trimmed.py
QUICK_COUNT_NON_CAAAAGCAG_TRIMMED_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/quick_count_non_CAAAAGCAG_trimmed.py

FASTQ_TO_TABLE = $(SCRIPTS_DIR)/fastq_to_table.py
FASTQ_TO_TABLE_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/fastq_to_table.py

PLOT_LENGTH_DISTRIBUTIONS = $(SCRIPTS_DIR)/plot_length_distributions.py
PLOT_LENGTH_DISTRIBUTIONS_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/plot_length_distributions.py

PSSM_PLOTS = $(SCRIPTS_DIR)/pssm_plots.py
PSSM_PLOTS_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/pssm_plots.py

U1_VS_U2 = $(SCRIPTS_DIR)/u1_vs_u2.py
U1_VS_U2_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/u1_vs_u2.py

LINEAR_REGRESSION = $(SCRIPTS_DIR)/linear_regression.py
LINEAR_REGRESSION_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/linear_regression.py

PLOT_PAR_LENGTHS = $(SCRIPTS_DIR)/plot_par_lengths.py
PLOT_PAR_LENGTHS_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/plot_par_lengths.py

FIND_DINUCLEOTIDE_AT_JUNCTION = $(SCRIPTS_DIR)/find_dinucleotide_at_junction.py
FIND_DINUCLEOTIDE_AT_JUNCTION_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/find_dinucleotide_at_junction.py

CONVERT_FORMATS = $(SCRIPTS_DIR)/convert_formats.py
CONVERT_FORMATS_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/convert_formats.py

PLOT_DISTS = $(SCRIPTS_DIR)/plot_dists.py
PLOT_DISTS_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/plot_dists.py

PLOT_EACH_LENGTH_DISTRIBUTION = $(SCRIPTS_DIR)/plot_each_length_distribution.py
PLOT_EACH_LENGTH_DISTRIBUTION_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/plot_each_length_distribution.py

PLOT_U_SNRNAS = $(SCRIPTS_DIR)/plot_u_snrnas.py
PLOT_U_SNRNAS_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/plot_u_snrnas.py

NOG_SCRIPT = $(SCRIPTS_DIR)/nog_script.py
NOG_SCRIPT_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/nog_script.py

NOG_STATS_COMBINER = $(SCRIPTS_DIR)/nog_stats_combiner.py
NOG_STATS_COMBINER_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/nog_stats_combiner.py

CONVERT_GTF_NAMES = $(SCRIPTS_DIR)/convert_gtf_names.py
CONVERT_GTF_NAMES_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/convert_gtf_names.py

TOP_GENES = $(SCRIPTS_DIR)/top_genes.py
TOP_GENES_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/top_genes.py

FIND_DINUCLEOTIDE_AT_JUNCTION_WITH_BACKGROUND = $(SCRIPTS_DIR)/find_dinucleotide_at_junction_with_background.py
FIND_DINUCLEOTIDE_AT_JUNCTION_WITH_BACKGROUND_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/find_dinucleotide_at_junction_with_background.py

TRIM_SEQUENCES = $(SCRIPTS_DIR)/trim_sequences.py
TRIM_SEQUENCES_LOCATION = $(INFLUENZA_SCRIPTS_DIR)/trim_sequences.py

# Link all the scripts!
$(foreach script,$(ALL_SCRIPTS),$(eval $(call link-file,$($(script)),$($(script)_LOCATION))))
